### DISCUSSION TITLE
Add ISource to handle  IReadOnlyDictionary<TKey, TValue>

### DISCUSSION THREAD

**Discussion by dombrovsky** [axuno/SmartFormat#342] [Q&A] [answered]
**Is your feature request related to a problem? Please describe.**
If arguments for formatting are coming as IReadOnlyDictionary (not upcastable to IDictionary), there is no built in Source that can handle it.
, 
**Describe the solution you'd like**
Should be a straightforward piece to add IReadOnlyDictionary support to DictionarySource, same way as it handles IDictionary.

**Describe alternatives you've considered**
Workaround is to convert IReadOnlyDictionary to IDictionary, but that would cause additional allocations.
There is always a possibility to implement custom Source on consumer side.

**Comment by axunonb** [comment, marked as answer]
You're saying that this tests fails for SmartFormat v3.2.1?

```CSharp
[Test]
public void ReadOnlyDictionaryTest()
{
    var dict = new Dictionary<string, object> { { "One", 1 }, { "Two", 2 }, { "Three", 3 }, };
    var roDict = new ReadOnlyDictionary<string, object>(dict);

    var formatter = Smart.CreateDefaultSmartFormat();
    var result = formatter.Format("{One}{Two}{Three}", roDict);

    Assert.That(result, Is.EqualTo("123"));
}
```

**Reply by dombrovsky** [reply]
No. I am saying it does not work if instance **would not be possible to cast to IDictionary<,>**. `ReadOnlyDictionary<,>` type implements both `IReadOnlyDictionary<,>` and `IDictionary<,>`.

So consider class like that:
```
internal sealed class MyReadOnlyDictionary<TKey, TValue> : IReadOnlyDictionary<TKey, TValue>
{
...
}
```
note it implements `IReadOnlyDictionary `and nothing else.

Changing your test so it has `var roDict = new MyReadOnlyDictionary<string, object>(dict);` would not work.

**Comment by axunonb** [comment]
```CSharp
public class CustomReadOnlyDictionary<TKey, TValue> : IReadOnlyDictionary<TKey, TValue>
{
    private readonly IDictionary<TKey, TValue> _dictionary;

    public CustomReadOnlyDictionary(IDictionary<TKey, TValue> dictionary)
    {
        _dictionary = dictionary;
    }

    public IEnumerator<KeyValuePair<TKey, TValue>> GetEnumerator()
    {
        return _dictionary.GetEnumerator();
    }

    IEnumerator IEnumerable.GetEnumerator()
    {
        return GetEnumerator();
    }

    public int Count => _dictionary.Count;

    public bool ContainsKey(TKey key)
    {
        return _dictionary.ContainsKey(key);
    }

    public bool TryGetValue(TKey key, out TValue value)
    {
        return _dictionary.TryGetValue(key, out value);
    }

    public TValue this[TKey key] => _dictionary[key];

    public IEnumerable<TKey> Keys => _dictionary.Keys;

    public IEnumerable<TValue> Values => _dictionary.Values;
}

var roDict = new CustomReadOnlyDictionary<string, object>(new Dictionary<string, object> { { "One", 1 }, { "Two", 2 }, { "Three", 3 }, });
var formatter = Smart.CreateDefaultSmartFormat();
var result = formatter.Format("{One}{Two}{Three}", roDict);

Assert.That(result, Is.EqualTo("123"));
```

This should succeed, but does not in any `Smart.Format` version because no `ISource` can handle ` IReadOnlyDictionary<TKey, TValue>`
